### DISCUSSION TITLE
How could you use HasOneThrough relationships?

### DISCUSSION THREAD

**Discussion by bluewhiskeywhisper** [rappasoft/laravel-livewire-tables#1357] [Q&A]
Hi there,

I'm trying to utilise a `HasOneThrough` relationship in my table. I have managed to get a `HasOne` relationship to work but I'm not able to get it working when applying a similar approach; I get this exception:

`Rappasoft\LaravelLivewireTables\Views\Column::setTable(): Argument #1 ($table) must be of type string, null given, called in Test_Project/vendor/rappasoft/laravel-livewire-tables/src/Traits/Helpers/ColumnHelpers.php on line 26`

I'm not quite sure what that means. My Post model has a relationship to Person via the Edit model:

    public function person(): HasOneThrough
    {
        return $this->hasOneThrough(Person::class, Edit::class, 'edit_id', 'id', 'id', 'person_id');
    }

I have pulled in the person relationship into my builder:

    public function builder(): Builder
    {
        return Post::query()
            ->with('person');
    }

I tried to reference `person.name` but it throws the above exception.

    public function columns(): array
    {
        return [
            Column::make("Person name", "person.name")
        ];
    }

I also have `$this->setEagerLoadAllRelationsStatus(true);`

Any help or suggestions would be greatly appreciated. Thank you. :)

**Using:** dev-v3-master (https://github.com/rappasoft/laravel-livewire-tables/commit/425051140b0dd711363116409821d7afba31a6ed), PHP 8.2.10, and Laveral 10.23.1.

**Comment by lrljoe** [comment]
Okay, let's do a couple of quick tests to see what's going on here shall we :)  I'm suspecting that some of your posts don't have a person.

You could do a withWhereHas if you **only** want to return Posts that have a Person associated, otherwise you'll need to either:

Use the built-in capability from Laravel to return a "default" value where a relation doesn't exist, or set some null values using the approach below:

```
            Column::make('Person name')
            ->label(
                fn ($row, Column $column) => (!is_null($row->person) ? $row->person->name : '')
            )
            ->sortable(),
```
Feel free to jump in the Discord, as I'm on there more often btw [https://discord.gg/mkmEfq7hUC](https://discord.gg/mkmEfq7hUC)

**Reply by bluewhiskeywhisper** [reply]
Thank you @lrljoe, yes, that worked but I'm not sure why. :)

I truncated my tables so that I have exactly 1 Post, which has exactly 2 Edits, each of which has exactly 1 Person. There are no null values. After doing this my code still failed; yours worked fine.

Do you suggest I continue with this approach, or does it indicate that something is not working correctly?

**Reply by lrljoe** [reply]
Comparing your code, and what you've said doesn't quite line up here.

Your code says that you have three models here

Post
Edit
Person

And a Post has One Person via the Edit model.

BUT

You've said that. post has two edits each with one person, in which case that relationship of has One is not valid in a Laravel sense, as, it has two persons (Many) via the edit model, not one.

Rather than trying to directly link the two, just pull the relation of the relation using the builder method.

What is your end goal for this particular table, you want to see the
Post (Title etc), and what else?

**Reply by bluewhiskeywhisper** [reply]
Hi @lrljoe, Yes, you're quite right; that would be a one-to-many relationship but I only want the latest to appear in my table, so I've just used a `hasOneThrough` relationship, which seems to work ok when I call it directly.

In terms of my end goal, here's what I'm hoping to achieve:

    +----+-----------+---------+
    | ID | Title     | Editor  | 
    +----+-----------+---------+
    | 1  | My test 1 | My name | 
    +----+-----------+---------+

Thanks for your help! :)

**Reply by lrljoe** [reply]
Ah, I know there a few oddities with latestOfMany etc.  

Let me review what I did last time to help fix it.

The interim solution of course would be to just have a last_edited_by or similar on your model, which is actually a more efficient approach

**Reply by bluewhiskeywhisper** [reply]
That's great, thank you @lrljoe!

Yes, I accept this might be more efficient in this case but if I can get this to work then I'll be extending the table to include lots of other related columns, which won't all be selecting the most recently added/created/etc

**Reply by lrljoe** [reply]
Did you solve the issue?  With the package finally getting to official beta etc, I may have some time to review a better solution.
